library(circlize)
library(doBy)
library(data.table)
library(ggplot2)
library(cowplot)
library(RColorBrewer)

# Takes as input 1Mbp or 100kbp files generated by:
# calculate_gencov_windows.R, calculate_heterozygosityDiff_windows.R
# Produces a circlized plot over each chromosome/scaffold and plots genome coverage 
# against difference in heterozygosity
# USE WITH CAUTION OUTSIDE PIPELINE
# Call: Rscript {r-script} {gencov nm=first} {gencov nm=second} {gencov nm=third} 
# {difference in heterozygosity} {circlized out} {scatterplot out} {chr-list}

source("code/functions.R")

set.seed(999)

args <- commandArgs(trailingOnly = TRUE)

file1 = args[1]
file2 = args[2]
file3 = args[3]
filesnp = args[4]
circlize_out = args[5]
scatter_out = args[6]
chr_file = args[7]

################################################################################
################################# READ FILES ###################################
################################################################################

cov_1_table <- read.table(file1, header=TRUE,fill=TRUE,stringsAsFactor=FALSE)
cov_2_table <- read.table(file2, header=TRUE,fill=TRUE,stringsAsFactor=FALSE)
cov_3_table <- read.table(file3, header=TRUE,fill=TRUE,stringsAsFactor=FALSE)
snp_table <- read.table(filesnp, header=TRUE,fill=TRUE,stringsAsFactor=FALSE)

if (dim(cov_1_table)[1] == 0) {
  
  print("Warning: No chromosome/scaffold over 1Mbp or 11kbp! Check output based on 
        chromosome instead.")
  
} else {
  
  if (file.exists(chr_file)) {
    
    chromosome <- read.csv(chr_file, header = FALSE, sep = ",")
    chromosome <- as.factor(chromosome)
    
    cov_1_table <- cov_1_table[cov_1_table$chr %in% chromosome, ]
    
  }
  
  nr_factors <- c(length(unique(cov_1_table$chr)),length(unique(cov_2_table$chr)),
                  length(unique(cov_3_table$chr)),length(unique(snp_table$chr)))
  
  # Makes sure that all tables have the same chromosomes
  while (min(nr_factors) != max(nr_factors)) {
    
    nr_factors_min <- which.min(nr_factors)
    
    if (nr_factors_min == 1) {
      
      cov_2_table <- cov_2_table[cov_2_table$chr %in% unique(cov_1_table$chr), ]
      cov_3_table <- cov_3_table[cov_3_table$chr %in% unique(cov_1_table$chr), ]
      snp_table <- snp_table[snp_table$chr %in% unique(cov_1_table$chr), ]
      
    } else if (nr_factors_min == 2) {
      
      cov_1_table <- cov_1_table[cov_1_table$chr %in% unique(cov_2_table$chr), ]
      cov_3_table <- cov_3_table[cov_3_table$chr %in% unique(cov_2_table$chr), ]
      snp_table <- snp_table[snp_table$chr %in% unique(cov_2_table$chr), ]
      
    } else if (nr_factors_min == 3) {
      
      cov_1_table <- cov_1_table[cov_1_table$chr %in% unique(cov_3_table$chr), ]
      cov_2_table <- cov_2_table[cov_2_table$chr %in% unique(cov_3_table$chr), ]
      snp_table <- snp_table[snp_table$chr %in% unique(cov_3_table$chr), ]
      
    } else if (nr_factors_min == 4) {
      
      cov_1_table <- cov_1_table[cov_1_table$chr %in% unique(snp_table$chr), ]
      cov_2_table <- cov_2_table[cov_2_table$chr %in% unique(snp_table$chr), ]
      cov_3_table <- cov_3_table[cov_3_table$chr %in% unique(snp_table$chr), ]
      
    }
    
    nr_factors <- c(length(unique(cov_1_table$chr)),length(unique(cov_2_table$chr)),
                    length(unique(cov_3_table$chr)),length(unique(snp_table$chr)))
    
  }
  
  
  cov_1 <- gen_data_4plotting(cov_1_table, c("chr", "range", "diff"))
  cov.select.1 <- cov_1$df
  max.cov.1 <- cov_1$max
  min.cov.1 <- cov_1$min
  median.cov.1 <- cov_1$median

  cov_2 <- gen_data_4plotting(cov_2_table, c("chr", "range", "diff"))
  cov.select.2 <- cov_2$df
  max.cov.2 <- cov_2$max
  min.cov.2 <- cov_2$min
  median.cov.2 <- cov_2$median

  cov_3 <- gen_data_4plotting(cov_3_table, c("chr", "range", "diff"))
  cov.select.3 <- cov_3$df
  max.cov.3 <- cov_3$max
  min.cov.3 <- cov_3$min
  median.cov.3 <- cov_3$median

  snp <- gen_data_4plotting(snp_table, c("chr", "range", "diff"))
  snp.select <- snp$df
  max.snp <- snp$max
  min.snp <- snp$min
  median.snp <- snp$median

################################################################################
############################# ORDER CHROMOSOMES ################################
################################################################################

  if (!file.exists(chr_file)) {
    
    # Order after scaffold length if no chromosome file is given
    max_per_chr <- setDT(cov.select.1)[, .SD[which.max(x)], by=factor]
    max_per_chr <- as.data.frame(max_per_chr[,1:2])
    chromosome <- max_per_chr[order(-max_per_chr$x),][,1]
    
  }
  
  cov.select.1$factor <- ordered(cov.select.1$factor, levels = chromosome)
  cov.select.2$factor <- ordered(cov.select.2$factor, levels = chromosome)
  cov.select.3$factor <- ordered(cov.select.3$factor, levels = chromosome)
  snp.select$factor <- ordered(snp.select$factor, levels = chromosome)
  
  factor.nr <- nr_factors[1]
  
################################################################################
################################ SCATTER PLOT ##################################
################################################################################

  cov.select <- merge(cov.select.1, cov.select.2, by = c("factor", "x"))
  colnames(cov.select) <- c("factor", "x", "cov1", "cov2")
  
  cov.select <- merge(cov.select, cov.select.3, by = c("factor", "x"))
  colnames(cov.select) <- c("factor", "x", "cov1", "cov2", "cov3")
  
  cov.select <- merge(cov.select, snp.select, by = c("factor", "x"))
  colnames(cov.select) <- c("Chromosomes", "window", "cov1", "cov2", "cov3", "hetDiff")
  
  
  # Get colors and shapes for all chromosomes
  point_colors <- c(brewer.pal(n = 8, name = "Dark2"))
  point_reps <- ceiling(factor.nr/length(point_colors))
  point_colors <- rep(point_colors, point_reps)[1:factor.nr]
  
  point_shapes <- c(rep(0,8),rep(1,8),rep(2,8),rep(3,8),
                    rep(4,8),rep(5,8),rep(6,8),rep(7,8))
  point_reps <- ceiling(factor.nr/length(point_shapes))
  point_shapes <- rep(point_shapes, point_reps)[1:factor.nr]
  
  
  cov1_plot <- ggplot(cov.select, aes(x = cov1, y = hetDiff)) + 
		            geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
                scale_color_manual(values = point_colors) +
                scale_shape_manual(values = point_shapes) +
                theme(legend.key.size = unit(2, "mm")) +
                labs(title = file1, x = "difference in normalized genome coverage",
                     y = "difference in heterozygosity") + 
                theme_bw()

  legend <- get_legend(cov1_plot)

  cov1_plot <- cov1_plot +
                theme(legend.position="none")
  
  cov2_plot <- ggplot(cov.select, aes(x = cov2, y = hetDiff)) + 
                geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
                scale_color_manual(values = point_colors) + 
                scale_shape_manual(values = point_shapes) +
                labs(title = file2, x = "difference in normalized genome coverage",
                y = "difference in heterozygosity") + 
                theme_bw() +
                theme(legend.position="none")
  
  cov3_plot <- ggplot(cov.select, aes(x = cov3, y = hetDiff)) + 
                geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
                scale_color_manual(values = point_colors) + 
                scale_shape_manual(values = point_shapes) +
                labs(title = file3, x = "difference in normalized genome coverage",
                y = "difference in heterozygosity") + 
                theme_bw() +
                theme(legend.position="none")

  p <- plot_grid(cov1_plot, cov2_plot, cov3_plot, nrow = 3, 
            labels = c("A","B","C"))
  
  l <- plot_grid(legend)
  
################################################################################
################################ SCATTER PLOT ##################################
################################################################################
  
  # Get mean and standard deviation for all chromosomes
  cov1_chr_sd = summaryBy(data=cov.select, cov1 ~ Chromosomes, FUN = sd)
  cov1_chr_mean = summaryBy(data=cov.select, cov1 ~ Chromosomes, FUN = mean)
  
  cov2_chr_sd = summaryBy(data=cov.select, cov2 ~ Chromosomes, FUN = sd)
  cov2_chr_mean = summaryBy(data=cov.select, cov2 ~ Chromosomes, FUN = mean)
  
  cov3_chr_sd = summaryBy(data=cov.select, cov3 ~ Chromosomes, FUN = sd)
  cov3_chr_mean = summaryBy(data=cov.select, cov3 ~ Chromosomes, FUN = mean)
  
  hetDiff_chr_sd = summaryBy(data=cov.select, hetDiff ~ Chromosomes, FUN = sd)
  hetDiff_chr_mean = summaryBy(data=cov.select, hetDiff ~ Chromosomes, FUN = mean)
  
  # Combine all statistics
  chr.stats = merge(cov1_chr_mean, cov1_chr_sd)
  chr.stats = merge(chr.stats, cov2_chr_mean)
  chr.stats = merge(chr.stats, cov2_chr_sd)
  chr.stats = merge(chr.stats, cov3_chr_mean)
  chr.stats = merge(chr.stats, cov3_chr_sd)
  chr.stats = merge(chr.stats, hetDiff_chr_mean)
  chr.stats = merge(chr.stats, hetDiff_chr_sd)
  
  
  cov1_plot <- ggplot(chr.stats, aes(x = cov1.mean, 
                                      xmin = cov1.mean - cov1.sd,
                                      xmax = cov1.mean + cov1.sd,
                                      y = hetDiff.mean, 
                                      ymin = hetDiff.mean - hetDiff.sd,
                                      ymax = hetDiff.mean + hetDiff.sd,)) +
    geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes)) + 
    geom_errorbarh(aes(color = Chromosomes)) +
    labs(title = file1, x = "difference in normalized genome coverage",
         y = "difference in heterozygosity") + 
    theme_bw() +
    theme(legend.position="none")
  
  cov2_plot <- ggplot(chr.stats, aes(x = cov2.mean, 
                                      xmin = cov2.mean - cov2.sd,
                                      xmax = cov2.mean + cov2.sd,
                                      y = hetDiff.mean, 
                                      ymin = hetDiff.mean - hetDiff.sd,
                                      ymax = hetDiff.mean + hetDiff.sd,)) +
    geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes)) + 
    geom_errorbarh(aes(color = Chromosomes)) +
    labs(title = file2, x = "difference in normalized genome coverage",
         y = "difference in heterozygosity") + 
    theme_bw() +
    theme(legend.position="none")
  
  cov3_plot <- ggplot(chr.stats, aes(x = cov3.mean, 
                                      xmin = cov3.mean - cov3.sd,
                                      xmax = cov3.mean + cov3.sd,
                                      y = hetDiff.mean, 
                                      ymin = hetDiff.mean - hetDiff.sd,
                                      ymax = hetDiff.mean + hetDiff.sd,)) +
    geom_point(aes(color = Chromosomes, shape = Chromosomes)) + 
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes)) + 
    geom_errorbarh(aes(color = Chromosomes)) +
    labs(title = file3, x = "difference in normalized genome coverage",
         y = "difference in heterozygosity") + 
    theme_bw() +
    theme(legend.position="none")
  
  c <- plot_grid(cov1_plot, cov2_plot, cov3_plot, nrow = 3, 
                 labels = c("A","B","C"))
  
  
  pdf(file=scatter_out, width = 9, height = 9)
  print(p)
  print(c)
  print(l)
  dev.off()

################################################################################
################################ CIRCLIZE PLOT #################################
################################################################################
  
  circos.clear()
  pdf(file=circlize_out, width = 9, height = 9)
  par(mfrow=c(1,1), mar = c(1, 1, 1, 1), lwd = 0.1, cex = 0.7) 
  circos.par(cell.padding = c(0, 0, 0, 0), "track.height" = 0.15, gap.after = c(rep(1, factor.nr-1), 10))
  circos.initialize(factors = cov.select.3$factor, x = cov.select.3$x)


  circos.trackPlotRegion(factors = snp.select$factor, ylim = c(-1,1), 
                         x = snp.select$x, y = snp.select$y, panel.fun = function(x, y) {
  
    grey = c("#FFFFFF", "#CCCCCC", "#999999")
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.text(mean(xlim), mean(ylim) + uy(9, "mm"), cex = 1.1, sector.index) 
    circos.lines(x, y, col = "blue", area = TRUE, baseline = 0)
    circos.yaxis(side = "left", sector.index = levels(cov.select.1$factor)[1])
  })

  circos.trackPlotRegion(factors = cov.select.1$factor, ylim = c(-1,1), 
                         x = cov.select.1$x, y = cov.select.1$y, panel.fun = function(x, y) {
  
    grey = c("#FFFFFF", "#CCCCCC", "#999999")
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.lines(x, y, col = "red", area = TRUE, baseline = 0)
    circos.yaxis(side = "left", sector.index = levels(cov.select.1$factor)[1])
  })

  circos.trackPlotRegion(factors = cov.select.2$factor, ylim = c(-1,1), 
                         x = cov.select.2$x, y = cov.select.2$y, panel.fun = function(x, y) {
  
    grey = c("#FFFFFF", "#CCCCCC", "#999999")
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.lines(x, y, col = "red", area = TRUE, baseline = 0)
    circos.yaxis(side = "left", sector.index = levels(cov.select.1$factor)[1])
  })

  circos.trackPlotRegion(factors = cov.select.3$factor, ylim = c(-1,1), 
                         x = cov.select.3$x, y = cov.select.3$y, panel.fun = function(x, y) {
  
    grey = c("#FFFFFF", "#CCCCCC", "#999999")
    sector.index = get.cell.meta.data("sector.index")
    xlim = get.cell.meta.data("xlim")
    ylim = get.cell.meta.data("ylim")
    circos.lines(x, y, col = "red", area = TRUE, baseline = 0)
    circos.yaxis(side = "left", sector.index = levels(cov.select.1$factor)[1])
    circos.axis("bottom", direction = "inside", labels.facing = "reverse.clockwise")
  })

  dev.off()

}

