library(doBy)
library(data.table)
library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(plyr)
library(gridGraphics)
library(plotly)
library(dplyr)
library(tidyverse)
library(ggpubr)
# Takes as input 1Mbp or 100kbp files generated by:
# calculate_gencov_windows.R, calculate_heterozygosityDiff_windows.R
# Produces a circlized plot over each chromosome/scaffold and plots genome 
# coverage against difference in heterozygosity
# USE WITH CAUTION OUTSIDE PIPELINE
# Call: Rscript {r-script} {gencov nm=first} {gencov nm=second} {gencov nm=third} 
# {difference in heterozygosity} {circlized out} {scatterplot out} {chr-list}

set.seed(999)

source("code/functions.R")
args <- commandArgs(trailingOnly = TRUE)

file1 = args[1]
file2 = args[2]
file3 = args[3]
filesnp = args[4]
absolute_out = args[5]
diff_out = args[6]
chr_file = args[7]
ED1 = args[8]
ED2 = args[9]
ED3 = args[10]
CHR_NR = args[11]
WINDOW = args[12]

#minRange = 800000
#setwd("~/Dropbox (MEEL)/PhD/Subprojects/6 Sylvioidea neosc-scan/9 Sexchromoscanner/code/")
#file1 = "../results/bella/aloatta/aloatta.HS.gencov.nodup.nm.0.0.1Mbp.out"
#file2 = "../results/bella/aloatta/aloatta.HS.gencov.nodup.nm.0.2.1Mbp.out"
#file3 = "../results/bella/aloatta/aloatta.HS.gencov.nodup.nm.all.1Mbp.out"
#filesnp = "../results/bella/aloatta/aloatta.HS.diffHeterozygosity.1Mbp.out"
#circlize_out = "test.circlize.pdf"
#scatter_out = "test.scatter.pdf"
#chr_file = "chr.txt"
#source("functions.R")
#ED1 = "0.0"
#ED2 = "0.2"
#ED3 = "all"
#CHR_NR = 5
#highlight_file = "highlight.txt"



ED1 = gsub("\\.", "-", ED1)
ED2 = gsub("\\.", "-", ED2)
ED3 = gsub("\\.", "-", ED3)
ED1 = gsub("$", " mismatches", ED1)
ED2 = gsub("$", " mismatches", ED2)


ED1 = gsub("0-0", "0", ED1)
ED1 = gsub("0-", "<= ", ED1)
ED2 = gsub("0-", "<= ", ED2)


CHR_NR <- as.integer(CHR_NR)

file1_base = gsub(".out$", "", file1)
file2_base = gsub(".out$", "", file2)
file3_base = gsub(".out$", "", file3)
filesnp_base = gsub(".out$", "", filesnp)

diff_out_base = gsub("\\.pdf", ".png", diff_out)
absolute_out_base = gsub("\\.pdf", ".png", absolute_out)

################################################################################
################################# READ FILES ###################################
################################################################################

cov_1_table <- read.table(file1, 
                          header=TRUE, 
                          fill=TRUE, 
                          stringsAsFactor=FALSE)

cov_2_table <- read.table(file2, 
                          header=TRUE, 
                          fill=TRUE, 
                          stringsAsFactor=FALSE)

cov_3_table <- read.table(file3, 
                          header=TRUE, 
                          fill=TRUE, 
                          stringsAsFactor=FALSE)

snp_table   <- read.table(filesnp, 
                          header=TRUE, 
                          fill=TRUE, 
                          stringsAsFactor=FALSE)

# Mean and standard deviation calculation


if ( dim( cov_1_table )[1] == 0) {
  
  print("Warning: No chromosome/scaffold over 1Mbp or 100kbp! Check output based 
        on chromosome instead.")
  
} else {
  
  fun <- function(x){
    c(m=mean(x), s = sd(x))
  }
  
  sd_cov_1_table <- summaryBy(diff ~ 1 , data = cov_1_table, FUN = fun)
  sd_cov_2_table <- summaryBy(diff ~ 1 , data = cov_2_table, FUN = fun)
  sd_cov_3_table <- summaryBy(diff ~ 1 , data = cov_3_table, FUN = fun)
  sd_snp_table <- summaryBy(diff ~ 1 , data = snp_table, FUN = fun)
  
  if ( file.exists( chr_file ) ) {
    
    chromosome <- read.delim(chr_file, 
                             header = FALSE)
    chromosome$V1 <- trimws(chromosome$V1, which = c("both", "left", "right"), whitespace = "[ \t\r\n]")
    chromosome <- chromosome$V1
    cov_1_table <- cov_1_table[ cov_1_table$chr %in% chromosome, ]
    cov_2_table <- cov_2_table[ cov_2_table$chr %in% chromosome, ]
    cov_3_table <- cov_3_table[ cov_3_table$chr %in% chromosome, ]
    snp_table <- snp_table[ snp_table$chr %in% chromosome, ]
    
  }
}


cov_1_table$diff_genome_mean <- sd_cov_1_table$diff.m
cov_1_table$diff_genome_sd <- sd_cov_1_table$diff.s
cov_2_table$diff_genome_mean <- sd_cov_2_table$diff.m
cov_2_table$diff_genome_sd <- sd_cov_2_table$diff.s
cov_3_table$diff_genome_mean <- sd_cov_3_table$diff.m
cov_3_table$diff_genome_sd <- sd_cov_3_table$diff.s
snp_table$diff_genome_mean <- sd_snp_table$diff.m
snp_table$diff_genome_sd <- sd_snp_table$diff.s

# Subset files for outlier (outside 95% CI) values
outlier.cov_1_table <- subset(cov_1_table, c(cov_1_table$diff > (cov_1_table$diff_genome_mean + cov_1_table$diff_genome_sd*2) | cov_1_table$diff < (cov_1_table$diff_genome_mean - cov_1_table$diff_genome_sd*2)))
outlier.cov_2_table <- subset(cov_2_table, c(cov_2_table$diff > (cov_2_table$diff_genome_mean + cov_2_table$diff_genome_sd*2) | cov_2_table$diff < (cov_2_table$diff_genome_mean - cov_2_table$diff_genome_sd*2)))
outlier.cov_3_table <- subset(cov_3_table, c(cov_3_table$diff > (cov_3_table$diff_genome_mean + cov_3_table$diff_genome_sd*2) | cov_3_table$diff < (cov_3_table$diff_genome_mean - cov_3_table$diff_genome_sd*2)))
outlier.snp_table <- subset(snp_table, c(snp_table$diff > (snp_table$diff_genome_mean + snp_table$diff_genome_sd*2) | snp_table$diff < (snp_table$diff_genome_mean - snp_table$diff_genome_sd*2)))

# Write outlier windows to tables
outname1 <- sprintf("%s.outlier.out", file1_base)
write.table(outlier.cov_1_table, file = outname1, sep = "\t", quote = FALSE, row.names = F)
outname2 <- sprintf("%s.outlier.out", file2_base)
write.table(outlier.cov_2_table, file = outname2, sep = "\t", quote = FALSE, row.names = F)
outname3 <- sprintf("%s.outlier.out", file3_base)
write.table(outlier.cov_3_table, file = outname3, sep = "\t", quote = FALSE, row.names = F)
outname4 <- sprintf("%s.outlier.out", filesnp_base)
write.table(outlier.snp_table, file = outname4, sep = "\t", quote = FALSE, row.names = F)


################################################################################
############################# ORDER CHROMOSOMES ################################
################################################################################

alldata <- rbind(cov_1_table, cov_2_table, cov_3_table, snp_table)


# Order after scaffold length if no chromosome file is given
if ( !file.exists(chr_file) ) {
  max_per_chr <- setDT( alldata )[, .SD[ which.max(range) ], 
                                  by=chr]
  max_per_chr <- as.data.frame( max_per_chr[,1:2] )
  chromosome <- max_per_chr[ order( -max_per_chr$range ), ][,1]
  len_chr <- dim(max_per_chr)[1]
} else { len_chr <- length(chromosome)

}


cov_1_table$chr <- ordered(cov_1_table$chr, 
                           levels = chromosome)
cov_2_table$chr <- ordered(cov_2_table$chr, 
                           levels = chromosome)
cov_3_table$chr <- ordered(cov_3_table$chr, 
                           levels = chromosome)
snp_table$chr <- ordered(snp_table$chr, 
                         levels = chromosome)


################################################################################
############################# FILTER MISSING ROWS ##############################
################################################################################

cov1_range <- cov_1_table[c(1,2)]
cov2_range <- cov_2_table[c(1,2)]
cov3_range <- cov_3_table[c(1,2)]
snp_range <- snp_table[c(1,2)]


ok_ranges <- merge(cov1_range, cov2_range, by =c("chr", "range"))
ok_ranges <- merge(ok_ranges, cov3_range, by =c("chr", "range"))
ok_ranges <- merge(ok_ranges, snp_range, by =c("chr", "range"))

cov_1_table <- merge(cov_1_table, ok_ranges, by =c("chr", "range"))
cov_2_table <- merge(cov_2_table, ok_ranges, by =c("chr", "range"))
cov_3_table <- merge(cov_3_table, ok_ranges, by =c("chr", "range"))
snp_table <- merge(snp_table, ok_ranges, by =c("chr", "range"))

cov_1_table$chr <- ordered(cov_1_table$chr, 
                           levels = chromosome)
cov_2_table$chr <- ordered(cov_2_table$chr, 
                           levels = chromosome)
cov_3_table$chr <- ordered(cov_3_table$chr, 
                           levels = chromosome)
snp_table$chr <- ordered(snp_table$chr, 
                         levels = chromosome)

################################################################################
############################### MANHATTAN PLOT #################################
################################################################################

data2 <- ggdraw() + draw_label(paste0("Data points from tables: \n", file1, " \n ", file2, " \n ", file3, " \n ", filesnp), hjust = 0.5)
data1 <- ggdraw() + draw_label(paste0("Chromosomes/scaffolds are ordered from longest to shortest (", CHR_NR, " longest chromosomes/scaffolds), \n unless a list of chromosomes/scaffolds is provided in the config file. \n If a list is provided, the chromosomes/scaffolds are ordered according to this list. \n Chromosome file: ", chr_file))
data3 <- ggdraw() + draw_label(paste0("Data points outside the 95% CI are in these tables: \n", outname1, "\n", outname2, "\n", outname3, "\n", outname4)) 

data <- plot_grid(data1, data2, data3, ncol = 1, rel_heights = c(1, 1, 1))
data2 <- plot_grid(data1, data2, ncol = 1, rel_heights = c(1, 1))



# Filter data for n longest scaffolds, determined by args[11]
if ( !file.exists(chr_file) & (len_chr > CHR_NR)) {
  nr_chr <- CHR_NR
  chromosome <- as.factor(chromosome[1:nr_chr])
} 

cov_1_table <- cov_1_table[cov_1_table$chr %in% chromosome,]
cov_2_table <- cov_2_table[cov_2_table$chr %in% chromosome,]
cov_3_table <- cov_3_table[cov_3_table$chr %in% chromosome,]
snp_table <- snp_table[snp_table$chr %in% chromosome,]


cov_1_table$heterogametic <- as.numeric(cov_1_table$heterogametic)
cov_1_table$homogametic <- as.numeric(cov_1_table$homogametic)
# Got code from here: https://www.r-graph-gallery.com/101_Manhattan_plot.html


colors <- c("heterogametic" = "dodgerblue", "homogametic" = "darkmagenta")

text_size_colour = list(theme_bw(base_family="Courier", base_size = 12) + 
    theme(axis.text.x= element_text(colour="black", size=12)) +
    theme(axis.text.y= element_text(colour="black", size=12)) +
    theme(axis.title.x = element_text(colour="black",size=14)) + 
    theme(axis.title.y = element_text(colour="black",size=14)) + 
    theme(plot.title=element_text(family="Courier", size=14, colour="black", hjust = 0.5)) +
    theme(plot.margin= margin(1, 1, 3, 7, "mm")) +
    theme(panel.border = element_rect(colour = "black", fill=NA, size=0.5)) +
    theme(panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "lightgrey"), 
          panel.grid.minor = element_line(size = 0.15, linetype = 'solid',
                                colour = "lightgrey")))

# Prepare the dataset
cov1 <- cov_1_table %>% 
  # Compute chromosome size
  group_by(chr) %>% 
  summarise(chr_len=max(range)) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(cov_1_table, ., by=c("chr"="chr")) %>%
  # Add a cumulative position of each SNP
  arrange(chr, range) %>%
  mutate( BPcum=range+tot)
# Prepare X axis
axisdf <- cov1 %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

maxX <- max(cov1$BPcum)

# Make the plot
p.cov1 <- ggplot(cov1, aes(x=BPcum, y=diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(0, 2)) +
  ylab(sprintf("genome coverage \n (%s)", ED1)) +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_point( aes(y = heterogametic, color="heterogametic"), alpha=0.5,size = 1 ) +
  geom_point( aes(y = homogametic, color="homogametic"), alpha=0.5,size = 1 ) +
  geom_smooth( aes(y = heterogametic, color="heterogametic", group = chr), method = 'gam', span = 0.3) +
  geom_smooth( aes(y = homogametic, color="homogametic", group = chr), method = 'gam', span = 0.3) +
  scale_color_manual(values = colors) +
  # Custom the theme:
  text_size_colour +
  theme( 
    legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1)) + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())


# Prepare the dataset
cov2 <- cov_2_table %>% 
  # Compute chromosome size
  group_by(chr) %>% 
  summarise(chr_len=max(range)) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(cov_2_table, ., by=c("chr"="chr")) %>%
  # Add a cumulative position of each SNP
  arrange(chr, range) %>%
  mutate( BPcum=range+tot)
# Prepare X axis
axisdf <- cov2 %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

maxX <- max(cov2$BPcum)

# Make the plot
p.cov2 <- ggplot(cov2, aes(x=BPcum, y=diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(0, 2)) + 
  ylab(sprintf("genome coverage \n (%s)", ED2)) +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_point( aes(y = heterogametic, color="heterogametic"), alpha=0.5,size = 1 ) +
  geom_point( aes(y = homogametic, color="homogametic"), alpha=0.5,size = 1 ) +
  geom_smooth( aes(y = heterogametic, color="heterogametic", group = chr), method = 'gam', span = 0.3) +
  geom_smooth( aes(y = homogametic, color="homogametic", group = chr), method = 'gam', span = 0.3) +
  scale_color_manual(values = colors) +
  # Custom the theme:
  text_size_colour +
  theme( 
    legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1))  + 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())


# Prepare the dataset
cov3 <- cov_3_table %>% 
  # Compute chromosome size
  group_by(chr) %>% 
  summarise(chr_len=max(range)) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(cov_3_table, ., by=c("chr"="chr")) %>%
  # Add a cumulative position of each SNP
  arrange(chr, range) %>%
  mutate( BPcum=range+tot)
# Prepare X axis
axisdf <- cov3 %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )

maxX <- max(cov3$BPcum)

# Make the plot
p.cov3 <- ggplot(cov3, aes(x=BPcum, y=diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center,guide = guide_axis(check.overlap = TRUE) ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(0, 2)) +
  ylab(sprintf("genome coverage \n (%s)", ED3)) +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_point( aes(y = heterogametic, color="heterogametic"), alpha=0.5,size = 1 ) +
  geom_point( aes(y = homogametic, color="homogametic"), alpha=0.5,size = 1 ) +
  geom_smooth( aes(y = heterogametic, color="heterogametic", group = chr), method = 'gam', span = 0.3) +
  geom_smooth( aes(y = homogametic, color="homogametic", group = chr), method = 'gam', span = 0.3) +
  scale_color_manual(values = colors) +
  # Custom the theme:
  text_size_colour +
  theme( 
    legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1)) 



# Prepare the dataset
snp <- snp_table %>% 
  # Compute chromosome size
  group_by(chr) %>% 
  summarise(chr_len=max(range)) %>% 
  # Calculate cumulative position of each chromosome
  mutate(tot=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>%
  # Add this info to the initial dataset
  left_join(snp_table, ., by=c("chr"="chr")) %>%
  # Add a cumulative position of each SNP
  arrange(chr, range) %>%
  mutate( BPcum=range+tot)
# Prepare X axis
axisdf <- snp %>% group_by(chr) %>% summarize(center=( max(BPcum) + min(BPcum) ) / 2 )


MaxDiff <- max(abs(pmax(snp$homogametic, snp$heterogametic)))
MinDiff <- MaxDiff - MaxDiff*2

maxX <- max(snp$BPcum)

# Make the plot
p.snp <- ggplot(snp, aes(x=BPcum, y=diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(MinDiff, MaxDiff)) +
  ylab("heterozygosity \n ") +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_point( aes(y = heterogametic, color="heterogametic"), alpha=0.5,size = 1 ) +
  geom_point( aes(y = homogametic, color="homogametic"), alpha=0.5,size = 1 ) +
  geom_smooth( aes(y = heterogametic, color="heterogametic", group = chr), method = 'gam', span = 0.3) +
  geom_smooth( aes(y = homogametic, color="homogametic", group = chr), method = 'gam', span = 0.3) +
  scale_color_manual(values = colors) +
  # Custom the theme:
  text_size_colour + 
  theme( 
    # legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1)) +
  theme(legend.title = element_blank()) + 
  theme(legend.text = element_text(colour="black", size=14)) +
  theme(legend.key.height= unit(1, 'cm'),legend.key.width= unit(1, 'cm'), legend.key=element_rect(fill='white'))+
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())


#legend_b <- get_legend(p.snp + guides(color = guide_legend(nrow = 1)) + theme_bw() + theme(legend.position = "bottom")) 

legend_b <- get_legend(p.snp + theme(legend.position = "bottom", legend.key=element_rect(fill='white')))

p.snp <- p.snp +  theme( 
  legend.position="none")

title <- ggdraw() + draw_label(sprintf("Genome-wide values (sexes plotted separately) across %s bp windows ", WINDOW), fontface='bold')

c <- ggarrange(p.snp, p.cov1, p.cov2, p.cov3, ncol=1, align = "hv",  labels = c("A","B","C", "D"), heights = c(1,1,1,1))
c <- plot_grid(title, c, ncol = 1, rel_heights = c(.1, 1))

#c <- plot_grid(title,
#               p.snp + theme(legend.position="none"), 
#               p.cov1 + theme(legend.position="none"), 
#               p.cov2 + theme(legend.position="none"), 
#               p.cov3 + theme(legend.position="none"), 
#               nrow = 5, 
#               rel_heights = c(0.2, 1,1,1,1),            
#               labels = c(" ","A","B","C", "D"))


d <- plot_grid(c, legend_b, ncol = 1, rel_heights = c(1, .1))

#pdf(file=absolute_out, width = 15, height = 10)
pdf(file=absolute_out, width = 14, height = 9)
print(d)
print(data2)
dev.off()


#png(file=absolute_out_base, width = 1200, height = 800)
#print(d)
#dev.off()

#absolute_out_base2 = gsub("\\.pdf", "", absolute_out)
#ggsave(sprintf("%s.png", absolute_out_base2), plot = d, device = png(), width = 14, height = 10, dpi = 900)



p.cov3 <- p.cov3 + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


c <- ggarrange(p.snp, p.cov1, p.cov2, p.cov3, ncol=1, align = "v",  labels = c("A","B","C","D"), heights = c(1,1,1,2))
c <- plot_grid(title, c, ncol = 1, rel_heights = c(.1, 1))


d <- plot_grid(c, legend_b, ncol = 1, rel_heights = c(1, .1))

absolute_out_base2 = gsub("\\.pdf", "", absolute_out)

#pdf(file=absolute_out, width = 15, height = 10)
pdf(file=sprintf("%s.verticalXaxis.pdf", absolute_out_base2), width = 14, height = 12)
print(d)
print(data2)
dev.off()


#png(file=absolute_out_base, width = 1200, height = 800)
#print(d)
#dev.off()

#absolute_out_base2 = gsub("\\.pdf", "", absolute_out)
#ggsave(sprintf("%s.verticalXaxis.png", absolute_out_base2), plot = d, device = png(), width = 14, height = 12, dpi = 900)



############### DIFF


maxX <- max(cov1$BPcum)

# Make the plot
b <- c(sd_cov_1_table$diff.m-(sd_cov_1_table$diff.s*2), sd_cov_1_table$diff.m, (sd_cov_1_table$diff.m+sd_cov_1_table$diff.s*2))
p.cov1 <- ggplot(cov1, aes(x=BPcum, y=diff, color = diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(-1, 1)) +
  ylab(sprintf("genome coverage \n (%s)", ED1)) +
  labs(color = "95 % CI") +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_rect(aes(xmin=-Inf, xmax = Inf, ymin=diff_genome_mean-diff_genome_sd*2, ymax=diff_genome_mean+diff_genome_sd*2), size = 0.01, color = "black", fill="grey80", alpha = 0.01) + 
  geom_point(alpha=0.8, size=1.5) +
  scale_color_gradientn(limits = c(-1,1), colours=c("blue", "grey40", "red"),breaks=b, labels=format(b)) + 
  text_size_colour +
  theme( 
    # legend.position="none",
 #   panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1)
  ) + guides(fill = guide_colourbar(barwidth = 0.5, barheight = 10))+ 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

maxX <- max(cov2$BPcum)
# Make the plot
b <- c(sd_cov_2_table$diff.m-(sd_cov_2_table$diff.s*2), sd_cov_2_table$diff.m, (sd_cov_2_table$diff.m+sd_cov_2_table$diff.s*2))
p.cov2 <- ggplot(cov2, aes(x=BPcum, y=diff, color = diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(-1, 1)) +
  ylab(sprintf("genome coverage \n (%s)", ED2)) +
  labs(color = "95 % CI") +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_rect(aes(xmin=-Inf, xmax = Inf, ymin=diff_genome_mean-diff_genome_sd*2, ymax=diff_genome_mean+diff_genome_sd*2), size = 0.01, color = "black", fill="grey80", alpha = 0.01) + 
  geom_point(alpha=0.8, size=1.5) +
  scale_color_gradientn(limits = c(-1,1), colours=c("blue", "grey40", "red"),breaks=b, labels=format(b)) + 
  text_size_colour +
  theme( 
    # legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1)) + 
  guides(fill = guide_colourbar(barwidth = 0.5, barheight = 10))+ 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

maxX <- max(cov3$BPcum)
# Make the plot
b <- c(sd_cov_3_table$diff.m-(sd_cov_3_table$diff.s*2), sd_cov_3_table$diff.m, (sd_cov_3_table$diff.m+sd_cov_3_table$diff.s*2))
p.cov3 <- ggplot(cov3, aes(x=BPcum, y=diff, color = diff)) +
  # custom X axis:
  scale_x_continuous(label = axisdf$chr, breaks= axisdf$center,guide = guide_axis(check.overlap = TRUE) ) +
  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(-1, 1)) +
  ylab(sprintf("genome coverage \n (%s)", ED3)) +
  labs(color = "95 % CI") +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_rect(aes(xmin=-Inf, xmax = Inf, ymin=diff_genome_mean-diff_genome_sd*2, ymax=diff_genome_mean+diff_genome_sd*2), size = 0.01, color = "black", fill="grey80", alpha = 0.01) + 
  geom_point(alpha=0.8, size=1.5) +
  scale_color_gradientn(limits = c(-1,1), colours=c("blue", "grey40", "red"),breaks=b, labels=format(b)) + 
  text_size_colour +
  theme( 
    # legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1))+ 
    guides(fill = guide_colourbar(barwidth = 0.5, barheight = 10))


MaxDiff <- max(abs(snp$diff))
MinDiff <- MaxDiff - MaxDiff*2
MaxPos <- max(abs(snp$BPcum))
MinPos <- min(abs(snp$BPcum))

maxX <- max(snp$BPcum)
b <- c(sd_snp_table$diff.m-(sd_snp_table$diff.s*2), sd_snp_table$diff.m, (sd_snp_table$diff.m+sd_snp_table$diff.s*2))
# Make the plot
p.snp <- ggplot(snp, aes(x=BPcum, y=diff, color = diff)) +
  # custom X axis:
  scale_x_continuous( label = axisdf$chr, breaks= axisdf$center ) +
#  scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis
  coord_cartesian(ylim=c(MinDiff, MaxDiff)) +
  ylab("heterozygosity \n ") +
  labs(color = "95 % CI") +
  geom_vline(aes(xintercept = tot), lty = 2, size = 0.2) +
  geom_vline(aes(xintercept = maxX), lty = 2, size = 0.2) +
  geom_rect(aes(xmin=-Inf, xmax = Inf, ymin=diff_genome_mean-diff_genome_sd*2, ymax=diff_genome_mean+diff_genome_sd*2), size = 0.01, color = "black", fill="grey80", alpha = 0.01) + 
  geom_point(alpha=0.8, size=1.5) +
  scale_color_gradientn(limits = c(MinDiff,MaxDiff), colours=c("blue", "grey40", "red"),breaks=b, labels=format(b)) + 
  text_size_colour +
  theme( 
    # legend.position="none",
#    panel.border = element_blank(),
    axis.title.x = element_blank(),
    panel.grid.major.x = element_blank(),
    panel.grid.minor.x = element_blank(),
    axis.text.x = element_text(vjust = 0.5, hjust=1))+ 
    guides(fill = guide_colourbar(barwidth = 0.5, barheight = 10))+ 
  theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank())

title <- ggdraw() + draw_label(sprintf("Genome-wide sex differences (heterogametic-homogametic) \n across %s bp windows", WINDOW), fontface='bold')

c <- ggarrange(p.snp, p.cov1, p.cov2, p.cov3, ncol=1, align = "hv",  labels = c("A","B","C", "D"), heights = c(1,1,1,1))
c <- plot_grid(title, c, ncol = 1, rel_heights = c(.1, 1))


#c- ggarrange(p.snp, p.cov1, p.cov2, p.cov3, nrow = 4)
#d <- plot_grid(c, legend_b, ncol = 1, rel_heights = c(1, .1))

pdf(file=diff_out, width = 14, height = 9)
#pdf(file="test.pdf", width = 15, height = 9)
print(c)
print(data)
dev.off()

#png(file=diff_out_base, width = 1100, height = 800)
#print(c)
#print(data)
#dev.off()


#diff_out_base2 = gsub("\\.pdf", "", diff_out)
#ggsave(sprintf("%s.png", diff_out_base2), plot = c, device = png(), width = 14, height = 10, dpi = 900)



p.cov3 <- p.cov3 + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust=1))


c <- ggarrange(p.snp, p.cov1, p.cov2, p.cov3, ncol=1, align = "v",  labels = c("A","B","C","D"), heights = c(1,1,1,2))
c <- plot_grid(title, c, ncol = 1, rel_heights = c(.1, 1))


d <- plot_grid(c, legend_b, ncol = 1, rel_heights = c(1, .1))

diff_out_base2 = gsub("\\.pdf", "", diff_out)

#pdf(file=absolute_out, width = 15, height = 10)
pdf(file=sprintf("%s.verticalXaxis.pdf", diff_out_base2), width = 14, height = 12)
print(d)
print(data)
dev.off()


#png(file=absolute_out_base, width = 1200, height = 800)
#print(d)
#dev.off()

#diff_out_base2 = gsub("\\.pdf", "", diff_out)
#ggsave(sprintf("%s.verticalXaxis.png", diff_out_base2), plot = d, device = png(), width = 14, height = 12, dpi = 900)


