library(doBy)
library(data.table)
library(ggplot2)
library(cowplot)
library(RColorBrewer)
library(plyr)
library(dplyr)
library(ggpubr)

# Takes as input 1Mbp or 100kbp files generated by:
# calculate_gencov_windows.R, calculate_heterozygosityDiff_windows.R
# Produces a circlized plot over each chromosome/scaffold and plots genome 
# coverage against difference in heterozygosity
# USE WITH CAUTION OUTSIDE PIPELINE
# Call: Rscript {r-script} {gencov nm=first} {gencov nm=second} {gencov nm=third} 
# {difference in heterozygosity} {circlized out} {scatterplot out} {chr-list}

set.seed(999)

source("workflow/scripts/functions.R")
args <- commandArgs(trailingOnly = TRUE)

file1 = args[1]
file2 = args[2]
file3 = args[3]
filesnp = args[4]
scatter_out = args[5]
chr_file = args[6]
highlight_file = args[7]
ED1 = args[8]
ED2 = args[9]
ED3 = args[10]
WINDOW = args[11]

ED1 = gsub("\\.", "-", ED1)
ED2 = gsub("\\.", "-", ED2)
ED3 = gsub("\\.", "-", ED3)
ED1 = gsub("$", " mismatches", ED1)
ED2 = gsub("$", " mismatches", ED2)

ED1 = gsub("0-0", "0", ED1)
ED1 = gsub("0-", "<= ", ED1)
ED2 = gsub("0-", "<= ", ED2)

file1_base = gsub(".out$", "", file1)
file2_base = gsub(".out$", "", file2)
file3_base = gsub(".out$", "", file3)
filesnp_base = gsub(".out$", "", filesnp)

scatter_out_base = gsub(".pdf$", "", scatter_out)

################################################################################
################################# READ FILES ###################################
################################################################################

ED1 = gsub("\\.", "-", ED1)
ED2 = gsub("\\.", "-", ED2)
ED3 = gsub("\\.", "-", ED3)

cov_1_table <- read.table( file1, 
                           header=TRUE, 
                           fill=TRUE, 
                           stringsAsFactor=FALSE)

cov_2_table <- read.table( file2, 
                           header=TRUE, 
                           fill=TRUE, 
                           stringsAsFactor=FALSE)

cov_3_table <- read.table( file3, 
                           header=TRUE, 
                           fill=TRUE, 
                           stringsAsFactor=FALSE)

snp_table   <- read.table( filesnp, 
                           header=TRUE, 
                           fill=TRUE, 
                           stringsAsFactor=FALSE)

if ( dim( cov_1_table )[1] == 0) {
  
  print("Warning: No chromosome/scaffold longer than the specified window size! Check output based 
        on chromosome instead.")
  
} else {
  
  if ( file.exists( chr_file ) ) {
    
    #chromosome <- read.csv(chr_file, 
    #                       header = FALSE, 
    #                       sep = ",")
    chromosome <- read.delim(chr_file, header = FALSE)
    chromosome$V1 <- trimws(chromosome$V1, which = c("both", "left", "right"), whitespace = "[ \t\r\n]")
    chromosome <- chromosome$V1
    cov_1_table <- cov_1_table[ cov_1_table$chr %in% chromosome, ]
    
  }
  
  nr_factors <- c( length( unique( cov_1_table$chr ) ),
                   length( unique( cov_2_table$chr ) ),
                   length( unique( cov_3_table$chr ) ),
                   length( unique( snp_table$chr )) )
  
  # Makes sure that all tables have the same chromosomes
  while ( min( nr_factors ) != max( nr_factors ) ) {
    
    nr_factors_min <- which.min(nr_factors)
    
    if ( nr_factors_min == 1 ) {
      
      cov_2_table <- cov_2_table[ cov_2_table$chr %in% unique(cov_1_table$chr), ]
      cov_3_table <- cov_3_table[ cov_3_table$chr %in% unique(cov_1_table$chr), ]
      snp_table <- snp_table[ snp_table$chr %in% unique(cov_1_table$chr), ]
      
    } else if ( nr_factors_min == 2 ) {
      
      cov_1_table <- cov_1_table[ cov_1_table$chr %in% unique(cov_2_table$chr), ]
      cov_3_table <- cov_3_table[ cov_3_table$chr %in% unique(cov_2_table$chr), ]
      snp_table <- snp_table[ snp_table$chr %in% unique(cov_2_table$chr), ]
      
    } else if ( nr_factors_min == 3 ) {
      
      cov_1_table <- cov_1_table[ cov_1_table$chr %in% unique(cov_3_table$chr), ]
      cov_2_table <- cov_2_table[ cov_2_table$chr %in% unique(cov_3_table$chr), ]
      snp_table <- snp_table[ snp_table$chr %in% unique(cov_3_table$chr), ]
      
    } else if ( nr_factors_min == 4 ) {
      
      cov_1_table <- cov_1_table[ cov_1_table$chr %in% unique(snp_table$chr), ]
      cov_2_table <- cov_2_table[ cov_2_table$chr %in% unique(snp_table$chr), ]
      cov_3_table <- cov_3_table[ cov_3_table$chr %in% unique(snp_table$chr), ]
      
    }
    
    nr_factors <- c( length(unique(cov_1_table$chr)),
                     length(unique(cov_2_table$chr)),
                     length(unique(cov_3_table$chr)),
                     length(unique(snp_table$chr)))
    
  }
  
  cov_1 <- gen_data_4plotting( cov_1_table, c("chr", "range", "diff"))
  cov.select.1 <- cov_1$df
  max.cov.1 <- cov_1$max
  min.cov.1 <- cov_1$min
  median.cov.1 <- cov_1$median
  
  cov_2 <- gen_data_4plotting( cov_2_table, c("chr", "range", "diff"))
  cov.select.2 <- cov_2$df
  max.cov.2 <- cov_2$max
  min.cov.2 <- cov_2$min
  median.cov.2 <- cov_2$median
  
  cov_3 <- gen_data_4plotting( cov_3_table, c("chr", "range", "diff"))
  cov.select.3 <- cov_3$df
  max.cov.3 <- cov_3$max
  min.cov.3 <- cov_3$min
  median.cov.3 <- cov_3$median
  
  snp <- gen_data_4plotting( snp_table, c("chr", "range", "diff"))
  snp.select <- snp$df
  max.snp <- snp$max
  min.snp <- snp$min
  median.snp <- snp$median
  
  ################################################################################
  ############################# ORDER CHROMOSOMES ################################
  ################################################################################
  
  # Order after scaffold length if no chromosome file is given
  if ( !file.exists(chr_file) ) {
    max_per_chr <- setDT( cov.select.1 )[, .SD[ which.max(x) ], 
                                         by=factor]
    max_per_chr <- as.data.frame( max_per_chr[,1:2] )
    chromosome <- max_per_chr[ order( -max_per_chr$x ), ][,1]
    
  }
  
  cov.select.1$factor <- ordered( cov.select.1$factor, 
                                  levels = chromosome)
  cov.select.2$factor <- ordered( cov.select.2$factor, 
                                  levels = chromosome)
  cov.select.3$factor <- ordered( cov.select.3$factor, 
                                  levels = chromosome)
  snp.select$factor <- ordered( snp.select$factor, 
                                levels = chromosome)
  
  ################################################################################
  ################################ SCATTER PLOT ##################################
  ################################################################################
  
  cov.select <- merge(cov.select.1, 
                      cov.select.2, 
                      by = c("factor", "x"))
  colnames(cov.select) <- c("factor", "x", "cov1", "cov2")
  
  cov.select <- merge(cov.select, 
                      cov.select.3, 
                      by = c("factor", "x"))
  colnames(cov.select) <- c("factor", "x", "cov1", "cov2", "cov3")
  
  cov.select <- merge(cov.select, 
                      snp.select, 
                      by = c("factor", "x"))
  colnames(cov.select) <- c("Chromosomes", "window", "cov1", "cov2", "cov3", "hetDiff")

  cov.select$highlight_col <- as.character(cov.select$Chromosomes)
  if ( file.exists( highlight_file ) ) {
    highlight <- read.delim(highlight_file, header = FALSE)
    autosome_list <- !(cov.select$highlight_col %in% highlight$V1)
    cov.select$highlight_col[autosome_list] <- "Autosomes"
    chr_list <- as.vector(highlight$V1)
    autosome_list <- "Autosomes"
    all_chr <- append(autosome_list, chr_list)
    cov.select <- cov.select[order(factor(cov.select$highlight_col, levels=unique(all_chr))),]
    
  }

  point_colors <- c("#A4A0A0",
    "#33a02c",               
    "#6a3d9a",
    "#e31a1c",
    "#1f78b4",
    "#b2df8a",
    "#cab2d6",
    "#ff7f00")
  # Get colors and shapes for all chromosomes
 # point_colors <- c( brewer.pal( n = 8, name = "Set1" ) )
  point_reps <- ceiling( nr_factors[1]/length( point_colors ) )
  point_colors <- rep( point_colors, point_reps )[1:nr_factors[1]]
  
  point_shapes <- c(rep(0,8),
                    rep(1,8),
                    rep(2,8),
                    rep(3,8),
                    rep(4,8),
                    rep(5,8),
                    rep(6,8),
                    rep(7,8))
  
  point_reps <- ceiling( nr_factors[1]/length( point_shapes ) )
  point_shapes <- rep( point_shapes, point_reps )[1:nr_factors[1]]

x_axis <- expression(Delta*~genome~coverage)
y_axis <- expression(Delta*~heterozygosity)

scaleFUN <- function(x) sprintf("%.2f", x)

text_size_colour = list(theme_bw(base_family="Courier", base_size = 12) + 
    theme(axis.text.x= element_text(colour="black", size=12)) +
    theme(axis.text.y= element_text(colour="black", size=12)) +
    theme(axis.title.x = element_text(colour="black",size=15)) + 
    theme(axis.title.y = element_text(colour="black",size=15)) + 
    theme(plot.title=element_text(family="Courier", size=15, colour="black", hjust = 0.5) ) +
    theme(panel.border = element_rect(colour = "black", fill=NA, size=0.5)) +
    theme(plot.margin= margin(1, 1, 1, 1, "mm")) +
    theme(panel.grid.major = element_line(size = 0.15, linetype = 'solid',
                                colour = "lightgrey"), 
          panel.grid.minor = element_line(size = 0.15, linetype = 'solid',
                                colour = "lightgrey"))) 

theme_title <- function(...) {
  theme_gray(base_family = "Courier") + 
    theme(plot.title = element_text(face = "bold"))
}
title_theme <- calc_element("plot.title", theme_title())

theme_description <- function(...) {
  theme_gray(base_family = "Courier") + 
    theme(plot.title = element_text(face = "plain",size = 12, hjust = 0))
}
theme_description <- calc_element("plot.title", theme_description())


  cov1_plot <- ggplot(cov.select, 
                      aes(x = cov1, 
                          y = hetDiff)) + 
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.1, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) + 

    scale_color_manual(values = point_colors) +
    scale_shape_manual(values = point_shapes) +
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +
    theme(legend.key.size = unit(2, "mm")) +
    labs(title = sprintf("%s", ED1), 
        x = " ", 
        y = y_axis) + 
    text_size_colour
  
  legend <- get_legend(cov1_plot)
  
  cov1_plot <- cov1_plot +
    theme(legend.position="none")
  
  cov2_plot <- ggplot(cov.select, 
                      aes(x = cov2, 
                          y = hetDiff)) + 
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.2, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) + 
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    labs(title = sprintf("%s", ED2), 
         x = x_axis,
         y = " ") + 
    text_size_colour +
    theme(legend.position="none")

  
  cov3_plot <- ggplot(cov.select, 
                      aes(x = cov3, 
                          y = hetDiff)) + 
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.3, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) + 
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    labs(title = sprintf("%s", ED3), 
         x = " ",
         y = " ") + 
    text_size_colour +
    theme(legend.position="none")

title <- ggdraw() + draw_label(sprintf("Sex differences (heterogametic-homogametic) per %s bp window", WINDOW), 
    fontfamily = title_theme$family,
    fontface = title_theme$face,
    size = title_theme$size
  )

    p <- ggarrange(cov1_plot, 
                   cov2_plot, 
                   cov3_plot, align="hv",
                   nrow = 1, 
                   labels = c("A","B","C"))

    p <- plot_grid(title, p, nrow = 2, rel_heights = c(0.15, 1))

    l <- plot_grid(legend)
  
  ################################################################################
  ################################ SCATTER PLOT ##################################
  ################################################################################
  
  # Get mean and standard deviation for all chromosomes
  cov1_chr_sd = summaryBy(data=cov.select, 
                          cov1 ~ Chromosomes + highlight_col, 
                          FUN = sd)
  cov1_chr_mean = summaryBy(data=cov.select, 
                            cov1 ~ Chromosomes + highlight_col, 
                            FUN = mean)
  
  cov2_chr_sd = summaryBy(data=cov.select, 
                          cov2 ~ Chromosomes + highlight_col, 
                          FUN = sd)
  cov2_chr_mean = summaryBy(data=cov.select, 
                            cov2 ~ Chromosomes + highlight_col, 
                            FUN = mean)
  
  cov3_chr_sd = summaryBy(data=cov.select, 
                          cov3 ~ Chromosomes + highlight_col, 
                          FUN = sd)
  cov3_chr_mean = summaryBy(data=cov.select, 
                            cov3 ~ Chromosomes + highlight_col, 
                            FUN = mean)
  
  hetDiff_chr_sd = summaryBy(data=cov.select, 
                             hetDiff ~ Chromosomes + highlight_col, 
                             FUN = sd)
  hetDiff_chr_mean = summaryBy(data=cov.select, 
                               hetDiff ~ Chromosomes + highlight_col, 
                               FUN = mean)
  
  chr_max = summaryBy(data=cov.select, 
                             window ~ Chromosomes + highlight_col, 
                             FUN = max)

  # Combine all statistics
  chr.stats = merge(cov1_chr_mean, 
                    cov1_chr_sd)
  chr.stats = merge(chr.stats, 
                    cov2_chr_mean)
  chr.stats = merge(chr.stats, 
                    cov2_chr_sd)
  chr.stats = merge(chr.stats, 
                    cov3_chr_mean)
  chr.stats = merge(chr.stats, 
                    cov3_chr_sd)
  chr.stats = merge(chr.stats, 
                    hetDiff_chr_mean)
  chr.stats = merge(chr.stats, 
                    hetDiff_chr_sd)
  chr.stats = merge(chr.stats, 
                    chr_max)
  
  chr.stats <- chr.stats[order(chr.stats$highlight_col),]
  chr.stats.A <- subset(chr.stats, chr.stats$highlight_col=="Autosomes")
  chr.stats.SC <- subset(chr.stats, chr.stats$highlight_col!="Autosomes")
  chr.stats <- rbind(chr.stats.A, chr.stats.SC)
  cov.select <- cov.select[order(cov.select$highlight_col),]
  cov.select.A <- subset(cov.select, cov.select$highlight_col=="Autosomes")
  cov.select.SC <- subset(cov.select, cov.select$highlight_col!="Autosomes")
  cov.select <- rbind(cov.select.A, cov.select.SC)
  
  x_range <- max(chr.stats$cov1.mean) - min(chr.stats$cov1.mean)
  y_range <- max(chr.stats$hetDiff.mean) - min(chr.stats$hetDiff.mean)

  cov1_plot <- ggplot(chr.stats, aes(x = cov1.mean, 
                                     xmin = cov1.mean - cov1.sd,
                                     xmax = cov1.mean + cov1.sd,
                                     y = hetDiff.mean, 
                                     ymin = hetDiff.mean - hetDiff.sd,
                                     ymax = hetDiff.mean + hetDiff.sd)) +
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.1, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) +
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +                 
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes), width = x_range/20) + 
    geom_errorbarh(aes(color = Chromosomes), height = y_range/20) +
    labs(title = sprintf("%s", ED1), 
         x = " ",
         y = y_axis) + 
    text_size_colour +
    theme(legend.position="none")
  
  x_range <- max(chr.stats$cov2.mean) - min(chr.stats$cov2.mean)
  cov2_plot <- ggplot(chr.stats, aes(x = cov2.mean, 
                                     xmin = cov2.mean - cov2.sd,
                                     xmax = cov2.mean + cov2.sd,
                                     y = hetDiff.mean, 
                                     ymin = hetDiff.mean - hetDiff.sd,
                                     ymax = hetDiff.mean + hetDiff.sd)) +
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.2, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) + 
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes), width = x_range/20) + 
    geom_errorbarh(aes(color = Chromosomes), height = y_range/20) +
    labs(title = sprintf("%s", ED2),
         x = x_axis,
         y = " ") + 
    text_size_colour +
    theme(legend.position="none")
  
  x_range <- max(chr.stats$cov3.mean) - min(chr.stats$cov3.mean)
  cov3_plot <- ggplot(chr.stats, aes(x = cov3.mean, 
                                     xmin = cov3.mean - cov3.sd,
                                     xmax = cov3.mean + cov3.sd,
                                     y = hetDiff.mean, 
                                     ymin = hetDiff.mean - hetDiff.sd,
                                     ymax = hetDiff.mean + hetDiff.sd)) +
    geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
    geom_vline(xintercept = median.cov.3, linetype="dashed", size=0.2) +
    geom_point(aes(color = Chromosomes, 
                   shape = Chromosomes), size = 3) + 
    scale_y_continuous(labels=scaleFUN) +
    scale_x_continuous(labels=scaleFUN) +
    scale_color_manual(values = point_colors) + 
    scale_shape_manual(values = point_shapes) +
    geom_errorbar(aes(color = Chromosomes), width = x_range/20) + 
    geom_errorbarh(aes(color = Chromosomes), height = y_range/20) +
    labs(title = sprintf("%s", ED3),
         x = " ",
         y = " ") + 
    text_size_colour +
    theme(legend.position="none")
  
title <- ggdraw() + draw_label(sprintf("Mean sex differences (heterogametic-homogametic) per chromosome/scaffold (across %s bp windows)", WINDOW),     
    fontfamily = title_theme$family,
    fontface = title_theme$face,
    size = title_theme$size)

    c <- ggarrange(cov1_plot, 
                   cov2_plot, 
                   cov3_plot, align="hv",
                   nrow = 1, 
                   labels = c("D","E","F"))


    c <- plot_grid(title, c, nrow = 2, rel_heights = c(0.15, 1))

  data <- ggdraw() + draw_label(paste0("Dashed lines are median values across all windows"),     
    fontfamily = theme_description$family,
    fontface = theme_description$face,
    size = theme_description$size)

  data2 <- ggdraw() + draw_label(paste0("Data points from tables: \n", file1, " \n ",
   file2, " \n ", file3, " \n ", filesnp),     
    fontfamily = theme_description$family,
    fontface = theme_description$face,
    size = theme_description$size)

  data <- plot_grid(data, data2, nrow = 2, rel_heights = c(1, 1))

  pdf(file=scatter_out, width = 12, height = 4.5)
  print(p)
  print(c)
  print(l)
  print(data)
  dev.off()
  
  if ( file.exists( highlight_file ) ) {
    cov.select$highlight_col <- factor(cov.select$highlight_col, levels = unique(cov.select$highlight_col))
    
        cov1_plot <- ggplot(cov.select, 
                        aes(x = cov1, 
                            y = hetDiff)) + 
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.1, linetype="dashed", size=0.2) +
      geom_point(aes(fill = highlight_col), pch = 21, size = 3) + 
      scale_fill_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      theme(legend.key.size = unit(2, "mm")) +
      labs(title = sprintf("%s", ED1), 
           x = " ",
           y = y_axis, fill="Chromosome type") + 
      text_size_colour
    
    legend <- get_legend(cov1_plot)
    
    cov1_plot <- cov1_plot +
      theme(legend.position="none")
    
    cov2_plot <- ggplot(cov.select, 
                        aes(x = cov2, 
                            y = hetDiff)) +
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.2, linetype="dashed", size=0.2) +                     
      geom_point(aes(fill = highlight_col), pch = 21, size = 3) + 
      scale_fill_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      labs(title = sprintf("%s", ED2), 
           x = x_axis,
           y = " ") + 
      text_size_colour +
      theme(legend.position="none")
    
    cov3_plot <- ggplot(cov.select, 
                        aes(x = cov3, 
                            y = hetDiff)) + 
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.3, linetype="dashed", size=0.2) +
      geom_point(aes(fill = highlight_col), pch = 21, size = 3) + 
      scale_fill_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      labs(title = sprintf("%s", ED3), 
           x = " ",
           y = " ") + 
      text_size_colour +
      theme(legend.position="none")
 
title <- ggdraw() + draw_label(sprintf("Sex differences (heterogametic-homogametic) per %s bp window", WINDOW),     
    fontfamily = title_theme$family,
    fontface = title_theme$face,
    size = title_theme$size)

    p <- ggarrange(cov1_plot, 
                   cov2_plot, 
                   cov3_plot, align="hv",
                   nrow = 1, 
                   labels = c("A","B","C"))

    p <- plot_grid(title, p, nrow = 2, rel_heights = c(0.15, 1))

    
    l <- plot_grid(legend)
    
    x_range <- max(chr.stats$cov1.mean) - min(chr.stats$cov1.mean)
    y_range <- max(chr.stats$hetDiff.mean) - min(chr.stats$hetDiff.mean)

    chr.stats$highlight_col <- factor(chr.stats$highlight_col, levels = unique(chr.stats$highlight_col))
    cov1_plot <- ggplot(chr.stats, aes(x = cov1.mean, 
                                       xmin = cov1.mean - cov1.sd,
                                       xmax = cov1.mean + cov1.sd,
                                       y = hetDiff.mean, 
                                       ymin = hetDiff.mean - hetDiff.sd,
                                       ymax = hetDiff.mean + hetDiff.sd)) +
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.1, linetype="dashed", size=0.2) +
      scale_fill_manual(values = point_colors) +
      scale_color_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      geom_errorbar(aes(color = highlight_col), width = x_range/20) + 
      geom_errorbarh(aes(color = highlight_col), height = y_range/20) +
      geom_point(aes(fill = highlight_col, size = window.max), pch = 21) + 
      labs(title = sprintf("%s", ED1), 
           x = " ",
           y = y_axis) + 
      text_size_colour +
      theme(legend.position="none") + 
      scale_size_continuous(range = c(2, 8))
    
    x_range <- max(chr.stats$cov2.mean) - min(chr.stats$cov2.mean)
    cov2_plot <- ggplot(chr.stats, aes(x = cov2.mean, 
                                       xmin = cov2.mean - cov2.sd,
                                       xmax = cov2.mean + cov2.sd,
                                       y = hetDiff.mean, 
                                       ymin = hetDiff.mean - hetDiff.sd,
                                       ymax = hetDiff.mean + hetDiff.sd)) +
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.2, linetype="dashed", size=0.2) +
      scale_fill_manual(values = point_colors) +
      scale_color_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      geom_errorbar(aes(color = highlight_col), width = x_range/20) + 
      geom_errorbarh(aes(color = highlight_col), height = y_range/20) +
      geom_point(aes(fill = highlight_col, size = window.max), pch = 21) + 
      labs(title = sprintf("%s", ED2),
           x = x_axis,
           y = " ") + 
      text_size_colour +
      theme(legend.position="none") + 
      scale_size_continuous(range = c(2, 8))
    

    x_range <- max(chr.stats$cov3.mean) - min(chr.stats$cov3.mean)
    cov3_plot <- ggplot(chr.stats, aes(x = cov3.mean, 
                                       xmin = cov3.mean - cov3.sd,
                                       xmax = cov3.mean + cov3.sd,
                                       y = hetDiff.mean, 
                                       ymin = hetDiff.mean - hetDiff.sd,
                                       ymax = hetDiff.mean + hetDiff.sd)) +
      geom_hline(yintercept = median.snp, linetype="dashed", size=0.2) +
      geom_vline(xintercept = median.cov.3, linetype="dashed", size=0.2) +
      scale_fill_manual(values = point_colors) +
      scale_color_manual(values = point_colors) +
      scale_y_continuous(labels=scaleFUN) +
      scale_x_continuous(labels=scaleFUN) +
      geom_errorbar(aes(color = highlight_col), width = x_range/20) + 
      geom_errorbarh(aes(color = highlight_col), height = y_range/20) +
      geom_point(aes(fill = highlight_col, size = window.max), pch = 21) + 
      labs(title = sprintf("%s", ED3),
           x = " ",
           y = " ") + 
      text_size_colour +
      theme(legend.position="none") + 
      scale_size_continuous(range = c(2, 8))

title <- ggdraw() + draw_label(sprintf("Mean sex differences (heterogametic-homogametic) per chromosome/scaffold (across %s bp windows)", WINDOW),     
    fontfamily = title_theme$family,
    fontface = title_theme$face,
    size = title_theme$size)

    c <- ggarrange(cov1_plot, 
                   cov2_plot, 
                   cov3_plot, 
                   nrow = 1, align="hv",
                   labels = c("D","E","F"))

    c <- plot_grid(title, c, nrow = 2, rel_heights = c(0.15, 1))

  data <- ggdraw() + draw_label(paste0("Dashed lines are median values across all windows"),     
    fontfamily = theme_description$family,
    fontface = theme_description$face,
    size = theme_description$size)

  data2 <- ggdraw() + draw_label(paste0("Data points from tables: \n", file1, " \n ",
   file2, " \n ", file3, " \n ", filesnp),     
    fontfamily = theme_description$family,
    fontface = theme_description$face,
    size = theme_description$size)
  dataColour <- ggdraw() + draw_label(paste0("Data points color from: \n", highlight_file),     
    fontfamily = theme_description$family,
    fontface = theme_description$face,
    size = theme_description$size)
  
  data <- plot_grid(data, 
                    data2,
                   dataColour, 
                   ncol = 1)

    outname <- sprintf("%s.highlight.pdf", scatter_out_base)
    pdf(file=outname, width = 12, height = 4.5)
    print(p)
    print(c)
    print(l)
    print(data)
    dev.off()  
    
  }
}
